### PR TITLE
[irods 7103] Add relative path support to iticket (main)

### PR DESCRIPTION
**Pull request by Nishant2022**
Update makeFullPath to actually return full path of object or collection

### CODE REVIEW

File: src/iticket.cpp [hunk 1]
```diff
@@ -457,18 +459,17 @@ showTickets( char *inName ) {
...
-
-std::string
-makeFullPath( const char *inName ) {
-    std::stringstream fullPathStream;
-    if ( strlen( inName ) == 0 ) {
-        return std::string();
-    }
-    if ( *inName != '/' ) {
-        fullPathStream << cwd << "/";
+auto makeFullPath(const char *inName) -> std::optional<std::string> {
```
Comment (korydraughn) [inline src/iticket.cpp:462]: This looks like the implementation from `istream` and other functions.

Let's just rename this function to match what's in `istream`. That way we know they are related and no one's confused.

**Comment by Nishant2022** [comment]
All unit tests passed. Most core tests passed. the `test_control_plane` test fails on a VM and skips on my local machine. We will look more into that.

**Comment by Nishant2022** [comment]
PR ready for review. All core tests pass apart from `tests_control_plane` (see https://github.com/irods/irods/pull/7153#issuecomment-1618982045). All unit tests pass.

File: src/iticket.cpp [hunk 2]
```diff
@@ -457,18 +459,17 @@ showTickets( char *inName ) {
...
-std::string
-makeFullPath( const char *inName ) {
-    std::stringstream fullPathStream;
-    if ( strlen( inName ) == 0 ) {
-        return std::string();
-    }
-    if ( *inName != '/' ) {
-        fullPathStream << cwd << "/";
+auto canonical(const char *inName) -> std::optional<std::string> {
+    char out_path[MAX_NAME_LEN]{};
```
Comment (alanking) [inline src/iticket.cpp:463]:
```suggestion
auto canonical(const char* inName) -> std::optional<std::string>
{
    char out_path[MAX_NAME_LEN]{};
```

**Review by alanking — COMMENTED** [review]
Looks good. Be sure to fix the commit message for the GitHub linkage, i.e.:

```
[irods/irods#7103] Add relative path support to iticket
```

**Comment by Nishant2022** [comment]
I updated the comment message for GitHub linkage and am ready to add the #'s.

File: src/iticket.cpp [hunk 3]
```diff
@@ -653,8 +655,13 @@ doCommand( char *cmdToken[] ) {
         else {
             makeTicket( myTicket );
         }
-        std::string fullPath = makeFullPath( cmdToken[2] );
-        doTicketOp( "create", myTicket, cmdToken[1], fullPath.c_str(),
+
+        auto fullPath = canonical(cmdToken[2]);
+        if (!fullPath.has_value()) {
+            return -2;
```
Comment (korydraughn) [inline src/iticket.cpp:661]: How did we arrive at `-2`?

Comment (alanking) [reply src/iticket.cpp:661]: My guess is because there's a -1 above this

Comment (korydraughn) [reply src/iticket.cpp:661]: I had a similar thought, but perhaps @Nishant2022 had a different reason.

Comment (Nishant2022) [reply src/iticket.cpp:661]: At the top of the current function there is a comment saying:

```c++
/* handle a command,
   return code is 0 if the command was (at least partially) valid,
   -1 for quitting,
   -2 for if invalid
   -3 if empty.
 */
``` 

If the canonical function does not return a value, then the provided path is invalid and the return value should reflect that.

Comment (korydraughn) [reply src/iticket.cpp:661]: Sounds good to me.

Comment (korydraughn) [reply src/iticket.cpp:661]: Please mark this as resolved.

**Comment by korydraughn** [comment]
Does the test pass with these changes?

**Comment by Nishant2022** [comment]
All test pass with these changes in addition to my test at https://github.com/irods/irods/pull/7166.

**Review by korydraughn — APPROVED** [review]
Excellent. Let's add the pound.

Like @alanking said, don't forget to include the `[irods/irods#7103]` as the commit message prefix.

**Comment by alanking** [comment]
Will merge this along with https://github.com/irods/irods/pull/7166.

**Pull request merged**